src/builtin.jq

```diff
@@ -6,9 +6,22 @@ def sort_by(f): _sort_by_impl(map([f]));
 def group_by(f): _group_by_impl(map([f]));
 def unique: group_by(.) | map(.[0]);
 def unique_by(f): group_by(f) | map(.[0]);
-def max_by(f): _max_by_impl(map([f]));
-def min_by(f): _min_by_impl(map([f]));
-def add: reduce .[] as $x (null; . + $x);
+
+def _minmax_by(gen; p; $ismin):
+    reduce (gen | [., p]) as $elem (null;
+      if (. == null) or (($elem[1] < .[1]) == $ismin) then $elem end
+    ) | .[0];
```

Comment:
I haven't checked but it would probably be a good idea to see what kind of performance impacts it is to do this in jq instead of C, i fear it's quite a lot slower. Maybe it's worth using separate implementation for array and generator?

Reply:
I personally prefer the elegance of a unified implementation (also, the jq version should at least have better memory usage since it doesn't create a temporary array containing all the `path_exp` results).

How about leaving the C implementation in (but commented out) for now? That way, this can be easily reverted if the performance impact turns out to affect anyone.